src/main/java/org/javarosa/xpath/expr/XPathExpression.java

```diff
@@ -76,6 +73,11 @@ private static void serializeElements(XPathNodeset nodeset, OutputStream output)
 
         DataInstance instance = nodeset.getInstance();
         Vector<TreeReference> refs = nodeset.getReferences();
+
+        if (refs == null) {
+            return;
```

Comment:
I don't think it's good practice for us to assume that the nodeset is well formed if refs is null here. What circumstance is that arising in? I think it's a stronger change for us to encode that situation deliberately rather than assuming that it was a valid state for the nodeset to have arrived in.